### PR TITLE
Emilce - RideShare - Octos

### PR DESCRIPTION
**Pull request by emilcecarlisa**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | Was struggling with how to implement a method that returns the completed/in-complete trips. Decided to make a helper method because it could be accessed in one spot instead of putting unless statements everywhere I wanted to total complete trips. 
Describe any examples of composition that you encountered in this project, if any |  My methods within a class are referencing one another. Also, the trip dispatcher is running methods from the passenger and driver classes. 
Describe a _nominal test_ that you wrote for this assignment. | For the aggregations, the drivers' total hourly revenue was tested. 
Describe an _edge case test_ that you wrote for this assignment. | I don't think I wrote any edge cases myself, unless you consider the test for nil end_times, ratings, and cost  as edge cases.
Describe a concept that you gained more clarity on as you worked on this assignment. | I'm starting to understand composition more, but I don't know if I could build up files like the ones provided to us.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -16,32 +17,60 @@ def initialize(input)
...
 
-      @trips = input[:trips] == nil ? [] : input[:trips]
+      @trips = input[:trips] == nil ? [] : input[:trips] #trips are objects in an array
+      @total_revenue = total_revenue
+    end
+
+    def completed_trips
+      @trips.find_all do |trip|
+        trip.end_time != nil && trip.rating != nil && trip.cost != nil
+      end
```
Comment (droberts-sea) [inline lib/driver.rb:29]: I love that you made this a helper method! Great job breaking the problem into small, reusable parts.

File: lib/passenger.rb [hunk 2]
```diff
@@ -13,12 +13,38 @@ def initialize(input)
...
+      end
+      return total
+    end
+
+    def total_time
+      total_time = 0
+
+      @trips.each do |trip|
+        total_time += trip.duration
+      end
```
Comment (droberts-sea) [inline lib/passenger.rb:45]: This should probably call `completed_trips`, otherwise it will produce an error.

File: lib/passenger.rb [hunk 3]
```diff
@@ -13,12 +13,38 @@ def initialize(input)
...
     end
 
     def add_trip(trip)
       @trips << trip
     end
+
+    def total_money
+      # total = @trips.inject {|sum, trip| sum + trip[@cost]} # why isn't this working!!!
+      total = 0
+
```
Comment (droberts-sea) [inline lib/passenger.rb:33]: On line 31, you're trying to use square brackets to get the value of an instance variable. This won't work - ruby will probably give you something like `Undefined Method '[]' for class Trip`. This is because square brackets only work to get values out of hashes and arrays.

When you've got an instance of a class, you need to use dot-notation to get a value: `trip.cost`

File: lib/trip_dispatcher.rb [hunk 4]
```diff
@@ -68,34 +68,72 @@ def load_trips
...
+      trip_id = (trips.last.id) + 1
+      available_driver = @drivers.find {|driver| driver.status == :AVAILABLE}
+      start_time =  Time.now
+      # if driver is available then return their id # && make them UNAVAILABLE
+      passenger = find_passenger(passenger_id)
+
+
+      data = {
+        id: trip_id,
+        driver: available_driver,
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:109]: I would probably call this something other than `data` - `raw_trip` or `trip_info` might be a more descriptive name.

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -58,23 +60,68 @@
...
-    it "returns zero if no trips" do
-      driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
-      driver.average_rating.must_equal 0
+    it "returns the total revenue for a trip" do
+      @driver.total_revenue.must_equal(12.04)
+    end
+
+    it "returns the average revenue per hour for a trip" do
+      @driver.avg_revenue.must_equal(0.07)
     end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:125]: You should break tests for different methods out into different `describe` blocks.

File: specs/driver_spec.rb [hunk 6]
```diff
@@ -58,23 +60,68 @@
...
+      driver.add_trip(trip2)
+      average = driver.average_rating
+      # Act/Assert
+      average.must_equal(5)
     end
+  end
+
+  describe "total revenue" do
+    before do
+      end_time = Time.parse("2016-05-02T09:56:00+00:00")
```
Comment (droberts-sea) [inline specs/driver_spec.rb:103]: Other questions that I would like to see these tests ask, for all the wave 1 methods (on both `Driver` and `Passenger`):

- What happens if the driver or passenger has no trips?
- What happens if the driver or passenger has an incomplete trip?

File: lib/trip_dispatcher.rb [hunk 7]
```diff
@@ -68,34 +68,72 @@ def load_trips
...
+
+        trip = Trip.new(data)
+        available_driver.add_trip(trip)
+        available_driver.status = :UNAVAILABLE
+        passenger.add_trip(trip)
+
+        @trips << trip # trip added to instance variable @trips
+
+        return data # nilclass being returned for the trip rating
+      end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:125]: Here you should be returning the instance of `Trip` in the `trip` variable, not the hash stored in `data`. They have all the same info, but the `Trip` is bundled into a neat package with methods you can call.

File: specs/trip_dispatch_spec.rb [hunk 8]
```diff
@@ -89,4 +89,19 @@
       passenger.trips.must_include trip
     end
   end
+
+  describe "adds new trips" do
+    before do
+      @dispatcher = RideShare::TripDispatcher.new
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:95]: There are a couple other things you should be testing around the `request_trip` method:

- Was the trip added to the trip lists of the driver and passenger?
- What if there is no available driver?

The second is an important case, and is exactly the sort of situation where software could do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Your commit messages look good, but I would like to see more frequent commits. If you're having trouble remembering to commit regularly, [you might find this article interesting](https://dev.to/rpalo/plan-your-commits)
Answer comprehension questions | yes
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | nominal cases are there but many other interesting tests are missing - see inline
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes - good work
Complete tests | some missing - see inline
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** |

This is a good start. Your code is clean and readable, and I really like the way you've demonstrated functional decomposition with the `completed_trips` method.

There are two areas where I see room for improvement. First, your test coverage could use some work. You're doing a good job of identifying nominal cases, but as you mentioned in the comprehension questions I'm not seeing any edge cases at all. You should work on finding the "edge" between working and not working, which means identifying the spots where a method you've written might do something funny. I've tried to outline some of these in the comments below.

Second, it seems like there's still some confusion around hashes and classes. In some places you returned a hash instead of an instance of a class, and in others I see you using hash notation to try and get values out of a class. [It might be worthwhile to review our lesson comparing the two](https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/01-ruby-fundamentals/hashes-classes-and-you.md), now that you've got a little more context.

That being said, your product code looks quite solid, and you've certainly solved the problem at hand. I've seen you putting in the hours, and I really appreciate that dedication - I'm excited to see what you come up with on Hotel. Keep up the hard work!
